public/app/features/explore/state/reducers.test.ts

```diff
@@ -231,6 +235,128 @@ describe('Explore item reducer', () => {
+          })
+        )
+        .thenStateShouldEqual(({
+          queries: [
+            { refId: 'A', key: 'mockKey' },
+            { refId: 'B', key: 'mockKey' },
+          ],
+          queryKeys: ['mockKey-0', 'mockKey-1'],
+        } as unknown) as ExploreItemState)
+        .whenActionIsDispatched(
+          removeQueryRowAction({
+            exploreId: ExploreId.left,
+            index: 0,
+          })
+        )
+        .thenStatePredicateShouldEqual((resultingState: ExploreItemState) => {
+          expect(resultingState.queries.length).toBe(1);
+          expect(resultingState.queries[0].refId).toBe('A');
+          expect(resultingState.queries[0].key).toMatch(QUERY_KEY_REGEX);
+          expect(resultingState.queryKeys[0]).toMatch(QUERY_KEY_REGEX);
+          return true;
+        });
+    });
+    it('reassigns query refId after removing a query to keep queries in order', () => {
+      reducerTester<ExploreItemState>()
+        .givenReducer(itemReducer, ({
+          queries: [],
+        } as unknown) as ExploreItemState)
+        .whenActionIsDispatched(
+          addQueryRowAction({
+            exploreId: ExploreId.left,
+            query: { refId: 'A' },
+            index: 0,
+          })
+        )
+        .thenStateShouldEqual(({
+          queries: [{ refId: 'A' }],
+          queryKeys: ['undefined-0'],
+        } as unknown) as ExploreItemState)
+        .whenActionIsDispatched(
+          addQueryRowAction({
+            exploreId: ExploreId.left,
+            query: { refId: 'B' },
+            index: 1,
+          })
+        )
+        .thenStateShouldEqual(({
+          queries: [{ refId: 'A' }, { refId: 'B' }],
+          queryKeys: ['undefined-0', 'undefined-1'],
+        } as unknown) as ExploreItemState)
+        .whenActionIsDispatched(
+          addQueryRowAction({
+            exploreId: ExploreId.left,
+            query: { refId: 'C' },
+            index: 2,
+          })
+        )
+        .thenStateShouldEqual(({
+          queries: [{ refId: 'A' }, { refId: 'B' }, { refId: 'C' }],
+          queryKeys: ['undefined-0', 'undefined-1', 'undefined-2'],
+        } as unknown) as ExploreItemState)
+        .whenActionIsDispatched(
+          removeQueryRowAction({
+            exploreId: ExploreId.left,
+            index: 0,
+          })
+        )
+        .thenStatePredicateShouldEqual((resultingState: ExploreItemState) => {
+          expect(resultingState.queries.length).toBe(2);
+          const queriesRefIds = resultingState.queries.map(query => query.refId);
+          const queriesKeys = resultingState.queries.map(query => query.key);
+          expect(queriesRefIds).toEqual(['A', 'B']);
+          queriesKeys.forEach(queryKey => {
+            expect(queryKey).toMatch(QUERY_KEY_REGEX);
+          });
+          resultingState.queryKeys.forEach(queryKey => {
+            expect(queryKey).toMatch(QUERY_KEY_REGEX);
+          });
+          return true;
+        });
```

Comment:
For easier maintainability and since I suppose this should focus on testing `removeQueryRowAction()` (you're testing `addQueryRowAction()` in another test already anyway, I'd remove all assertions except the final one.

And something that's in a similar vein but a bit more a matter of taste - also reduce the number of added query rows to 2 instead of 3, since `['B']` turning to `['A']` should be just as safe as `['B', 'C']` turning into `['A', 'B']`.